### ISSUE TITLE
apimachinery resource.Quantity primitive values should be public for recursive hashing

### ISSUE THREAD

**Issue by njtran** [kubernetes/kubernetes#124114]
### What would you like to be added?

API Machinery resource.Quantity{} has primitive values that represent the quantity of the resource it represents [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go#L101-L106), represented through private variables. Since these are private, not public, this prevents controllers that hash CRDs with libraries like https://github.com/mitchellh/hashstructure, which recursively walk through the structs, from properly identifying changes made to the quantity.

### Why is this needed?

In github.com/kubernetes-sigs/karpenter, we hash our CRDs and use hash comparisons to detect spec-Drift, similar to pods/deployments. We use this for fields like kubeReserved and systemReserved, which impacts the allocatable resources on bin-packing. If a user defines a different kubeReserved memory value, we want to be able to detect this as a change in our hashing/caching logic, but since it's private, we just have to wait for our cache values to expire before it's refreshed.

As a workaround, we're changing our v1.ResourceList type variables to type map[string]string where we then have to parse + validate each of the key value pairs as resource quantities, but it would be nice if this type of workflow was supported natively by making the values public, or at least making a public facing value that's generated from the private values.

Is there a reason these fields are private? Is this something that would be supported?

**Comment by alexzielenski** [comment]
I don't think it is advisable to hash the object internals. There may be fields unrelated to the hash/object value expressed, and best kept as an implementation detial. 

What you are doing now by hashing the string representation of the object is probably the best thing to do here.

/triage accepted
/cc @benluddy

**Comment by benluddy** [comment]
> a public facing value that's generated from the private values

To add to @alexzielenski's advice, hashing the serialized form of an object (e.g. using `encoding/json`) is ideal for this. It guarantees that every change you detect in the controller corresponds to a change in the representation of the object as ultimately seen by the API server.
